### PR TITLE
Add BitBox02 Simulator

### PR DESCRIPTION
**Pull request by asi345**
Recently, a simulator for BitBox02 was implemented. Therefore, we also wanted to add this to HWI repository for automated tests and ease of reference in the future.

### CODE REVIEW

**Comment by benma** [comment]
fyi @achow101, I reviewed this before @asi345 made the PR

File: test/test_device.py
Comment (achow101) [inline test/test_device.py:132]: `supports_legacy` should be a property of the emulator, not an argument to the test case. See `supports_taproot`.

File: test/test_device.py
Comment (achow101) [inline test/test_device.py:587]: It should be fine to just change the address type to `sh_wit` always instead of using `legacy` for the other devices.

File: test/setup_environment.sh
Comment (achow101) [inline test/setup_environment.sh:448]: This script needs to work when run inside a docker container as that is how the tests are suggested to be run locally. IIRC Docker within docker doesn't really work, so I don't think this will.

File: test/test_device.py
Comment (asi345) [reply test/test_device.py:132]: Okay, we can use this convention. I'll updating that shortly, but then I will also add the relevant field to the other emulators.

File: test/test_device.py
Comment (asi345) [reply test/test_device.py]: Okay, changed the types to `sh_wit`

**Comment by asi345** [comment]
Okay, we also came up with a solution for docker in docker problem for bitbox02 simulator.
Now, a separate `build_bitbox02.sh` script is ran before running hwi container, to build BitBox02 simulator.
See the comments in ci Dockerfile for reference.

File: test/README.md
Comment (benma) [inline test/README.md:356]: Replace with 

```
cd bitbox02-firmware
make dockerpull
```

(remove the `cd` below then)

File: test/setup_environment.sh
Comment (benma) [inline test/setup_environment.sh:445]: use `$(cat .containerversion)` instead of `latest` here too

**Comment by benma** [comment]
@achow101 please approve the workflow and take another look at the changes, would be really nice to have this merged soon. Thanks!

**Comment by achow101** [comment]
Could you rebase so that CI runs the tests with the simulator?

File: test/README.md
Comment (brunoerg) [inline test/README.md:342]: Isn't there a way to build it without docker?

Comment (benma) [reply test/README.md:342]: Not really, this is how we build and release the BB02 firmware. Replicating the Docker environment outside of it is bound to lead to tons of overhead in maintenance and CI breaks.

podman instead of docker would also work though.

Comment (brunoerg) [reply test/README.md:342]: Ok, anyway, I could build and run it. No problem.

File: test/test_bitbox02.py
Comment (brunoerg) [inline test/test_bitbox02.py:109]: I had some problems with this test. It seems you should pass bitcoind and interface in `bitbox02_test_suite ` instead of None, no?

File: test/setup_environment.sh
Comment (brunoerg) [inline test/setup_environment.sh:449]: Got this error running `setup_environment.sh`:
```
+ docker run -i --rm --volume ./:/bb02 shiftcrypto/firmware_v2:42 bash -c 'git config --global --add safe.directory /bb02 && cd /bb02 && make -j simulator'
docker: Error response from daemon: create ./: "./" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path.
```

I think you could use `"$(pwd):/bb02"` instead of `./:/bb02`.

File: test/test_bitbox02.py
Comment (asi345) [reply test/test_bitbox02.py:109]: Yes, you're right. Fixed that now.

File: test/setup_environment.sh
Comment (asi345) [reply test/setup_environment.sh:449]: This was not an issue in CI, but in local maybe that created problems. Anyways, I changed to `$(pwd)` instead of `.` as you suggested. This should solve the error.

**Comment by asi345** [comment]
Okay, I rebased this branch on top of master. CI is again waiting for your approval.

However, can you in any chance fix `bitcoind` build first? This commit breaks `bitcoind` build in `setup_environment.sh`, since `autogen.sh` file is not existent in bitcoin repo anymore : https://github.com/bitcoin/bitcoin/commit/d71ac768424333b65a6d88c9752cc9c7fdb276f3

**Comment by brunoerg** [comment]
> However, can you in any chance fix bitcoind build first? This commit breaks bitcoind build in setup_environment.sh, since autogen.sh file is not existent in bitcoin repo anymore : https://github.com/bitcoin/bitcoin/commit/d71ac768424333b65a6d88c9752cc9c7fdb276f3

See https://github.com/bitcoin-core/HWI/pull/752

File: ci/build_bitbox02.sh
Comment (brunoerg) [inline ci/build_bitbox02.sh:5]:
```suggestion
docker run -i --rm -v bitbox02_volume:/bitbox02-firmware shiftcrypto/firmware_v2:$CONTAINER_VERSION bash -c \
```

File: hwilib/devices/bitbox02.py
Comment (brunoerg) [inline hwilib/devices/bitbox02.py:270]: It throw`ConnectionRefusedError` if it wasn't able to connect to the BitBox02 simulator, so we won't be able to use any other simulator.

Comment (asi345) [reply hwilib/devices/bitbox02.py:270]: As far as I saw, Trezor or Keepkey also call `connect` on their simulators without handling any exceptions or `connect_ex`. Do you have a suggestion how to handle this exactly?

Comment (brunoerg) [reply hwilib/devices/bitbox02.py:270]: > As far as I saw, Trezor or Keepkey also call connect on their simulators without handling any exceptions or connect_ex. 

See:
```py
def enumerate(password: Optional[str] = None, expert: bool = False, chain: Chain = Chain.MAIN, allow_emulators: bool = False) -> List[Dict[str, Any]]:
    results = []
    devs = hid.HidTransport.enumerate()
    devs.extend(webusb.WebUsbTransport.enumerate())
    if allow_emulators:
        devs.extend(udp.UdpTransport.enumerate())
    for dev in devs:
```

When calling `enumerate` with `--simulators`, Trezor will call `enumerate` from `UdpTransport`. This function will return nothing if it wasn't able to find a device (simulator) so `devs` is empty.

Maybe you could do something like:
```diff
diff --git a/hwilib/devices/bitbox02.py b/hwilib/devices/bitbox02.py
index 53497c6..7cd6929 100644
--- a/hwilib/devices/bitbox02.py
+++ b/hwilib/devices/bitbox02.py
@@ -186,6 +186,8 @@ def enumerate(password: Optional[str] = None, expert: bool = False, chain: Chain
         devs.append(SIMULATOR_PATH)
     for path in devs:
         client = Bitbox02Client(path=path)
+        if allow_emulators and not client.transport:
+            continue
         if path != SIMULATOR_PATH:
             client.set_noise_config(SilentNoiseConfig())
         d_data: Dict[str, object] = {}
@@ -267,7 +269,11 @@ class BitBox02Simulator():
     def __init__(self) -> None:
         self.client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
         ip, port = SIMULATOR_PATH.split(":")
-        self.client_socket.connect((ip, int(port)))
+        self.connected = True
+        try:
+            self.client_socket.connect((ip, int(port)))
+        except:
+            self.connected = False
 
     def write(self, data: bytes) -> None:
         # Messages from client are always prefixed with HID report ID(0x00), which is not expected by the simulator.
@@ -301,7 +307,10 @@ class Bitbox02Client(HardwareWalletClient):
             self.noise_config = CLINoiseConfig()
         else:
             simulator = BitBox02Simulator()
-            self.transport = u2fhid.U2FHid(simulator)
+            if simulator.connected:
+                self.transport = u2fhid.U2FHid(simulator)
+            else:
+                self.transport = None
         self.device_path = path
 
         # use self.init() to access self.bb02.
```

**Comment by achow101** [comment]
The current implementation is causing all other tests to fail.

File: hwilib/devices/bitbox02.py
Comment (asi345) [reply hwilib/devices/bitbox02.py]: Okay, I did a fix based on this(not the same since this does not pass type checks). Now, in my tests, other devices are also passing the tests.

**Review by Thompson1985 — APPROVED** [review]

**Review by Thompson1985 — APPROVED** [review]

**Review by Thompson1985 — APPROVED** [review]

**Review by Thompson1985 — APPROVED** [review]

**Review by Thompson1985 — APPROVED** [review]

**Comment by benma** [comment]
> The current implementation is causing all other tests to fail.

Please take a look again, @asi345 updated the PR and all checks have passed.

**Comment by achow101** [comment]
ACK 3c3a185200fb92281128d0d84e0697da21f7dc09

**Pull request merged**
